pkg/kubelet/dockershim/remote/docker_server.go

```diff
@@ -0,0 +1,128 @@
+)
+
+const (
+	// defaultEndpoint is the default address of dockershim grpc server socket.
+	defaultAddress = "/var/run/dockershim.sock"
+	// unixProtocol is the network protocol of unix socket.
+	unixProtocol = "unix"
+	// startTimeout is the timeout to wait for dockershim grpc server to become ready.
+	startTimeout = 2 * time.Minute
+)
+
+// DockerServer is the grpc server of dockershim.
+type DockerServer struct {
+	// addr is the address to serve on.
+	addr string
+	// service is the docker service which implements runtime and image services.
+	service *DockerService
+	// listener is the network listener. It should be closed properly before the
+	// server is stopped.
+	listener net.Listener
+	// server is the grpc server.
+	server *grpc.Server
+}
+
+// NewDockerServer creates the dockershim grpc server.
+func NewDockerServer(addr string, s dockershim.DockerService) *DockerServer {
+	return &DockerServer{
+		addr:    addr,
+		service: NewDockerService(s),
+	}
+}
+
+// Start starts the dockershim grpc server.
+func (s *DockerServer) Start() error {
+	glog.V(2).Infof("Start dockershim grpc server")
+	// Unlink to cleanup the previous socket file.
+	err := syscall.Unlink(s.addr)
+	if err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("failed to unlink socket file %q: %v", s.addr, err)
+	}
+	l, err := net.Listen(unixProtocol, s.addr)
+	if err != nil {
+		return fmt.Errorf("failed to listen on %q: %v", s.addr, err)
+	}
+	s.listener = l
+	// Create the grpc server and register runtime and image services.
+	s.server = grpc.NewServer()
+	runtimeApi.RegisterRuntimeServiceServer(s.server, s.service)
+	runtimeApi.RegisterImageServiceServer(s.server, s.service)
+	errCh := make(chan error)
+	go func() {
+		defer close(errCh)
+		// Use interrupt handler to make sure the server to be stopped properly.
+		h := interrupt.New(nil, s.Stop)
+		err := h.Run(func() error { return s.server.Serve(l) })
+		if err != nil {
+			errCh <- fmt.Errorf("failed to serve connections: %v", err)
+		}
+	}()
+	// We need readinessCheck now because currently we start dockershim grpc server
+	// in-process, and will connect the server right after starting it. So it must
+	// be ready before the function returns.
+	return s.readinessCheck(errCh)
+}
+
+// readinessCheck checks the readiness of dockershim grpc server. Once there is an error
+// in errCh, the function will stop waiting and return the error.
+func (s *DockerServer) readinessCheck(errCh <-chan error) error {
+	glog.V(2).Infof("Running readiness check for docker server")
+	endTime := time.Now().Add(startTimeout)
+	for endTime.After(time.Now()) {
+		select {
+		case err := <-errCh:
+			return err
+		case <-time.After(time.Second):
+			glog.V(4).Infof("Connecting to runtime service %q", s.addr)
+			// The connection timeout doesn't matter here because
+			// we don't really use any remote runtime function here.
+			_, err := remote.NewRemoteRuntimeService(s.addr, 1*time.Minute)
```

Comment:
This method logs the error at `glog.Error` level, so this will spam the logs. That method should not log the error, since it is returned. I think grpc.Dial has a spammy log message too...